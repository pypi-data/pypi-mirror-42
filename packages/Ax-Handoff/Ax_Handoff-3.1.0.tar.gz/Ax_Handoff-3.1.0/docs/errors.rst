==============================================================================
Errors & Exceptions
==============================================================================

The ``axonchisel.handoff.error`` module contains all Exceptions raised by 
this package:
    
- ``SerializeError``:
    Error encoding an object to wire format.
        
- ``UnserializeError``
    Error decoding an object from wire format.

- ``DataTamperedError``
    Subclass of UnserializeError indicating data
    was tampered with or corrupted in transit.

