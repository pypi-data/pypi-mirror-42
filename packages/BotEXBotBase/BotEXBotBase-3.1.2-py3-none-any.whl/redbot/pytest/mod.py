import pytest

__all__ = ["mod"]


@pytest.fixture
def mod(config, monkeypatch):
    from redbot.core import Config

    with monkeypatch.context() as m:
        m.setattr(Config, "get_conf", lambda *args, **kwargs: config)
        from redbot.core import modlog

        modlog._register_defaults()
        return modlog
