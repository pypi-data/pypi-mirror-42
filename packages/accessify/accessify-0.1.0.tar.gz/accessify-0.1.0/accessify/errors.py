"""
Provide errors for accessibility levels.
"""
INACCESSIBLE_DUE_TO_ITS_PROTECTION_LEVEL_EXCEPTION_MESSAGE = \
    '{class_name}.{method_name}() is inaccessible due to its protection level'


class InaccessibleDueToItsProtectionLevelException(Exception):
    """
    Inaccessible due to its protection level exception.
    """

    def __init__(self, message):
        self.message = message