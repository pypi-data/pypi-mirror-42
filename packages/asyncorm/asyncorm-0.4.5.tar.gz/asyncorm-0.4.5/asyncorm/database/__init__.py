from asyncorm.database.db_manager import PostgresManager
from asyncorm.database.db_cursor import Cursor

__all__ = ["PostgresManager", "Cursor"]
