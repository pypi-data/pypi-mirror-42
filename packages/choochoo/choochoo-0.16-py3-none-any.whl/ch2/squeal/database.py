
from contextlib import contextmanager

from sqlalchemy import create_engine, event
from sqlalchemy.engine import Engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.sql.functions import count

from . import *
from .support import Base
from ..command.args import DATABASE, NamespaceWithVariables, NO_OP, parser
from ..lib.log import make_log

# mention these so they are "created"
Source,  Interval,
ActivityGroup, ActivityJournal, ActivityTimespan,
Topic, TopicJournal,
StatisticName, StatisticJournal, StatisticJournalInteger, StatisticJournalFloat, StatisticJournalText,
Segment, SegmentJournal
Pipeline
MonitorJournal
Constant, SystemConstant
ActivitySimilarity, ActivityNearby


# https://stackoverflow.com/questions/13712381/how-to-turn-on-pragma-foreign-keys-on-in-sqlalchemy-migration-script-or-conf
@event.listens_for(Engine, "connect")
def fk_pragma_on_connect(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA foreign_keys=ON;")  # https://www.sqlite.org/pragma.html#pragma_foreign_keys
    cursor.execute("PRAGMA temp_store=MEMORY;")  # https://www.sqlite.org/pragma.html#pragma_temp_store
    cursor.execute("PRAGMA threads=4;")  # https://www.sqlite.org/pragma.html#pragma_threads
    cursor.execute("PRAGMA cache_size=-1000000;")  # 1GB  https://www.sqlite.org/pragma.html#pragma_cache_size
    cursor.execute("PRAGMA secure_delete=OFF;")  # https://www.sqlite.org/pragma.html#pragma_secure_delete
    # cursor.execute("PRAGMA journal_mode=WAL;")  # https://www.sqlite.org/wal.html
    cursor.close()


@event.listens_for(Engine, 'close')
def analyxe_pragma_on_close(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA optimize;")  # https://www.sqlite.org/pragma.html#pragma_optimize
    cursor.close()


class Database:

    def __init__(self, args, log):
        self._log = log
        path = args.file(DATABASE)
        self._log.info('Using database at %s' % path)
        self.engine = create_engine('sqlite:///%s' % path, echo=False)
        self.session = sessionmaker(bind=self.engine)
        self.__create_tables()

    def __create_tables(self):
        if self.is_empty(tables=True):
            self._log.info('Creating tables')
            Base.metadata.create_all(self.engine)

    @contextmanager
    def session_context(self):
        session = self.session()
        try:
            yield session
            session.commit()
        except:
            session.rollback()
            raise
        finally:
            session.close()

    def is_empty(self, tables=False):
        if tables:
            # https://stackoverflow.com/questions/33053241/sqlalchemy-if-table-does-not-exist
            return not self.engine.dialect.has_table(self.engine, Source.__tablename__)
        else:
            with self.session_context() as s:
                n_topics = s.query(count(Topic.id)).scalar()
                n_activities = s.query(count(ActivityGroup.id)).scalar()
                n_statistics = s.query(count(StatisticName.id)).scalar()
            return not (n_topics + n_activities + n_statistics)


def add(s, instance):
    s.add(instance)
    return instance


def connect(args):
    '''
    Bootstrap from commandline-like args.
    '''
    p = parser()
    if len(args) == 1:
        args = args[0].split()
    elif args:
        args = list(args)
    else:
        args = []
    args.append(NO_OP)
    ns = NamespaceWithVariables(p.parse_args(args))
    log = make_log(ns)
    db = Database(ns, log)
    return ns, log, db

