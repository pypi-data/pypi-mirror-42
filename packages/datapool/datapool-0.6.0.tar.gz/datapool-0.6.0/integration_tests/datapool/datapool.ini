[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

