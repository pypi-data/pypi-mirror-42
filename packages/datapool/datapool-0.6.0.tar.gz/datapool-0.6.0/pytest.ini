[pytest]
addopts = -rs --full-trace
testpaths = tests
