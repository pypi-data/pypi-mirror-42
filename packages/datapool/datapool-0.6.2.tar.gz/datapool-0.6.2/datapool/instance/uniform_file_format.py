# encoding: utf-8
from __future__ import print_function, division, absolute_import

import csv
import datetime
import io
import os

from datapool.utils import iter_to_list


def _check_header(header):

    header = list(header)

    if "site" in header:
        header.remove("site")
    else:
        for n in "xyz":
            if n not in header:
                raise ValueError(
                    "header {!r} must have either 'site' or 'x', 'y' and 'z'".format(
                        header
                    )
                )
            header.remove(n)

    header = set(header)
    fix_column_names = set(("timestamp", "parameter", "value"))
    if header != fix_column_names:
        missing = ["'{}'".format(name) for name in fix_column_names - header]
        invalid = ["'{}'".format(name) for name in header - fix_column_names]
        msg = ""
        if missing:
            msg += "missing: {}".format(", ".join(sorted(missing)))
        if invalid:
            msg += "invalid: {}".format(", ".join(sorted(invalid)))
        raise ValueError("invalid header ({})".format(msg))


@iter_to_list
def read_from_file(path):
    if not os.path.exists(path):
        raise ValueError("{} does not exist".format(path))

    with open(path, "r", encoding="ascii") as fh:
        yield from _read_from_fh(fh)


@iter_to_list
def read_from_string(data):
    fh = io.StringIO(data)
    yield from _read_from_fh(fh)


def can_be_converted_to_float(value):
    try:
        value = float(value)
    except ValueError:
        return False
    return True


def parse_timestamp(timestamp):
    return datetime.datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S")


def _read_from_fh(fh):
    reader = csv.reader(fh, delimiter=";")
    header = [f.strip() for f in next(reader)]
    _check_header(header)
    for i, row in enumerate(reader):
        if len(row) < len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is incomplete".format(i + 1, row))
        if len(row) > len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is to long".format(i + 1, row))

        row = (cell.strip() for cell in row)
        row_dict = dict(zip(header, row))
        yield row_dict


@iter_to_list
def check_rows(row_dicts):

    if not row_dicts:
        return

    for i, row_dict in enumerate(row_dicts):

        value = row_dict["value"]
        if not can_be_converted_to_float(value):
            yield "row {}: value '{}' is not a float".format(i, value)
        else:
            row_dict["value"] = float(value)

        timestamp = row_dict["timestamp"]
        try:
            row_dict["timestamp"] = parse_timestamp(timestamp)
        except ValueError:
            yield "row {}: timstamp '{}' either has wrong format or is invalid".format(
                i, timestamp
            )

        has_xyz = all(name in row_dict and row_dict[name] is not None for name in "xyz")
        not_empty = ("parameter",)
        if not has_xyz:
            not_empty += ("site",)

        for name in not_empty:
            if row_dict[name].strip() == "":
                yield "row {}: field {} is empty".format(i, name)


@iter_to_list
def to_signals(row_dicts):

    for row in row_dicts:
        for f in "xyz":
            if f in row:
                row["coord_" + f] = row[f]
                del row[f]

    from .domain_objects import Signal

    return list(map(Signal, row_dicts))
