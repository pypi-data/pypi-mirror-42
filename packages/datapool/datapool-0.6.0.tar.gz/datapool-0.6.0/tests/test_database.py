# encoding: utf-8
from __future__ import absolute_import, division, print_function

import pytest

from datapool.database import (
    check_if_tables_exist,
    copy_db,
    dump_db,
    filters_to_sqlalchemy_expression,
    setup_db,
    setup_fresh_db,
)
from datapool.errors import InvalidOperationError
from datapool.instance.config_handling import config_for_develop_db
from datapool.instance.site_and_picture_model import check_and_commit

from .test_site_model import load_sites


def test_setup_db(config, setup_logger):

    with pytest.raises(InvalidOperationError):
        setup_fresh_db(config.db)

    setup_db(config.db)
    assert check_if_tables_exist(config.db)

    with pytest.raises(InvalidOperationError):
        setup_db(config.db)

    setup_fresh_db(config.db)


def test_copy_db(config, lz, regtest, setup_logger, tmpdir):

    engine = setup_db(config.db)
    excs, (site,) = load_sites(lz)
    assert not excs
    check_and_commit(site, engine)
    dump_db(config.db, file=regtest)

    config_develop_db, __ = config_for_develop_db(tmpdir.strpath)

    # the target table does not exist, but delete_existing=True should still work in
    # this case
    for name in copy_db(config.db, config_develop_db, delete_existing=True):
        print("copy", name, file=regtest)
    dump_db(config.db, file=regtest)

    # now we try to overwrite existing tables:
    with pytest.raises(InvalidOperationError):
        for __ in copy_db(config.db, config_develop_db, delete_existing=False):
            pass

    # now we enforce overwriting existing data:
    for __ in copy_db(config.db, config_develop_db, delete_existing=True):
        pass
    dump_db(config.db, file=regtest)


def test_filters_to_sqlalchemy_expression(regtest):
    all_filters = [
        ["timestamp > 2017, site==abc"],
        ["timestamp <= 2017, site==abc", "x==7"],
        ["timestamp > 2017, source==abc"],
        ["timestamp >= 2017, parameter==a"],
        ["timestamp, parameter==a"],
        ["timestamp != 2017, parameter==a"],
        ["timestamp == 2017-13, parameter==a"],
        ["timestamp == 2017-13, invalid_field==a"],
    ]

    for filters in all_filters:
        print(filters, file=regtest)
        expression, messages = filters_to_sqlalchemy_expression(filters)
        for message in messages:
            print(message, file=regtest)
        if expression is not None:
            print(expression, sorted(expression.compile().params.items()), file=regtest)
        print(file=regtest)
