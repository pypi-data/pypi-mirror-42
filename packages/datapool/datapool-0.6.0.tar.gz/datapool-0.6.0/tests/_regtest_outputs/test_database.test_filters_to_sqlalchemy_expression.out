['timestamp > 2017, site==abc']
signal.timestamp > :timestamp_1 AND site.name = :name_1 [('name_1', 'abc'), ('timestamp_1', datetime.datetime(2017, 1, 1, 0, 0))]

['timestamp <= 2017, site==abc', 'x==7']
signal.timestamp <= :timestamp_1 AND site.name = :name_1 AND signal.coord_x = :coord_x_1 [('coord_x_1', '7'), ('name_1', 'abc'), ('timestamp_1', datetime.datetime(2017, 1, 1, 0, 0))]

['timestamp > 2017, source==abc']
signal.timestamp > :timestamp_1 AND source.name = :name_1 [('name_1', 'abc'), ('timestamp_1', datetime.datetime(2017, 1, 1, 0, 0))]

['timestamp >= 2017, parameter==a']
signal.timestamp >= :timestamp_1 AND parameter.name = :name_1 [('name_1', 'a'), ('timestamp_1', datetime.datetime(2017, 1, 1, 0, 0))]

['timestamp, parameter==a']
filter timestamp has not the form FIELD CMP VALUE with CMP in (==, <=, <, >=, >)

['timestamp != 2017, parameter==a']
filter timestamp != 2017 has not the form FIELD CMP VALUE with CMP in (==, <=, <, >=, >)

['timestamp == 2017-13, parameter==a']
datetime 2017-13 is not of format '%Y-%m-%d:%H:%M:%S' or contains invalid values

['timestamp == 2017-13, invalid_field==a']
datetime 2017-13 is not of format '%Y-%m-%d:%H:%M:%S' or contains invalid values
filter invalid_field==a has invalid field invalid_field, allowed are parameter, site, source, timestamp, x, y, z

